semexprs: make `semAsgn` error node aware

zerbina opened the pull request:
## Summary

- make `asgnToResultVar` `nkError` aware
- remove input AST mutations from  `semAsgn`
- replace `localReport`s with `nkError` usage in `semAsgn`
- make `semAsgn` propagate `nkError`s

In addition, the `semIterator`, `semMethod`, and `semMacroDef` now return early when `semProcAux` returns an error, in which they would previously have crashed.

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented on compiler/sem/semexprs.nim:
```diff
@@ -2193,26 +2211,26 @@ proc semAsgn(c: PContext, n: PNode; mode=asgnNormal): PNode =
+            result[1] = typeMismatch(c.config, n.info, lhs.typ, rhsTyp, rhs)
+            hasError = true
+
+      if not hasError:
```
The old code and diff are a bit tricky to follow, but I _believe_ this has to guard on `if result[1].kind != nkError`, because otherwise we cut the rhs analysis short and that _might_ be why CI is failing. Just a guess.

saem reviewed (approved):
Looks good to me.

I read through the code to see if I could spot where the CI errors are originating from, I have one guess which I left as a comment.

zerbina commented:
Thanks for taking a look @saem. The tests are failing for three reasons:
1. I missed adjusting the `n.typ = c.enforceVoidContext` line, so assignments to `result` no longer enforce a void context
2. `nkError` nodes in routine bodies are reported twice, once by `trackProc` and once by `trackStmt` (I'll make a fix in a separate PR)
3. `semReturn` is not `nkError` aware (fixed by the PR that's already up)

zerbina committed:
semexprs: make `semAsgn` error node aware

Summary
=======

- make `asgnToResultVar` `nkError` aware
- remove input AST mutations from  `semAsgn`
- replace `localReport`s with `nkError` usage in `semAsgn`
- make `semAsgn` propagate `nkError`s

saem commented:
bors r+
